frontend/apps/reader/modules/readertoc.lua

```diff
@@ -215,6 +215,34 @@ function ReaderToc:getTocIndexByPage(pn_or_xp)
     return pre_index
 end
 
+function ReaderToc:getAccurateTocIndexByXPointer(xptr)
+    local pageno = self.ui.document:getPageFromXPointer(xptr)
+    local index = self:getTocIndexByPage(pageno) -- get toc entry on the page
+    local i = index
+    if not index or not self.toc[index] then return end
+    local initial_comparison = self.ui.document:compareXPointers(self.toc[index].xpointer, xptr)
+    if initial_comparison and initial_comparison < 0 then
+        while self.toc[i] do
+            local toc_xptr = self.toc[i].xpointer
+            local cmp = self.ui.document:compareXPointers(toc_xptr, xptr)
+            if cmp and cmp >= 0 then
+                return i
+            end
+            i = i + 1
```

Comment:
I can't access the previous commit where you were using `index`, but I think you want to scan backward here, so it should be `i = i - 1` until you find `i` where `cmp>=0`.
Also, please add a comment along `if cmp and cmp >= 0 then -- toc_xptr is before xptr` (and the opposite alongside the `else` so we don't have to go look at base/cre.cpp to know what a return valud of +1/0/-1 means :)